include/seastar/core/memory.hh

```diff
@@ -431,43 +435,59 @@ struct allocation_site {
     }
 };
 
-/// @brief If memory profiling is on returns the current memory live set
+/// @brief If memory sampling is on returns the current sampled memory live set
+///
+/// If there is tracked allocations (because heap profiling was on earlier)
+/// these will still be returned if heap profiling is now off
+///
 /// @return a vector of \ref allocation_site
-std::vector<allocation_site> memory_profile();
+std::vector<allocation_site> sampled_memory_profile();
 
-/// @brief Copies the current set of allocation_sites in the output parameter
+/// @brief Copies the current sampled set of allocation_sites in the output parameter
 /// (up to size of output vector)
 ///
 /// Returns amount of copied elements. Useful if one wants to avoid allocating
 /// an output vector (e.g.: under OOM conditions)
 ///
 /// @param output vector to copy the allocation sites to
 /// @return number of \ref allocation_site copied to the vector
-size_t memory_profile(std::vector<allocation_site>& output);
+size_t sampled_memory_profile(std::vector<allocation_site>& output);
 
-/// Enable heap profiling
+/// @brief Enable sampled heap profiling by setting a sample rate
+///
+/// @param sample_rate the sample rate to use. Disable heap profiling by setting
+/// the sample rate to 0
 ///
 /// In order to use heap profiling you have to define
 /// `SEASTAR_HEAPPROF`.
 ///
+/// Use \ref sampled_memory_profile for API access to profiling data
+///
+/// Note: Changing the sampling rate is currently not supported.
+/// Re-enabling heap profiling with a different sample rate to previous is fine
+/// to do if and only if all allocations are freeed before heap profiling is
+/// turned back on.
```

Comment:
Yeah possibly. What would you suggest if the check fails though? `abort()`?

Reply:
Abort, yes.

If we think aborting would be too dangerous/unexpected to the poor user who attempts to use it in production without testing it, then we could put a warning in the one place a user is guaranteed to read — the name of the function. Like `assert_there_are_no_allocations_and_set_heap_profiling_sampling_rate.`

@avikivity What's your opinion on this?

Reply:
Perhaps the comment also makes it sound scarier than it is. I should clarify that re-enabling sampling with a different rate doesn't cause any crash, corruption or undefined behavior: rather you are simply mixing samples with different frequencies and so your the aggregate samples will be biased, i.e., the samples which were taken when the sampling rate were lower should have a higher weight in principle but nothing like that is implement.

So the comment there from @StephanDollberg is indicating what you would need to do to avoid this bias (not have any allocations, which would also mean no samples), which of course is not really feasible in any case.

My recommendation would be that we soften the comment to detail the "problems" with mixing sampling of different rates but that we don't actually block it because it would still be useful in some situations. FWIW, the way we use the profiler currently in Redpanda is that is always on, and at a fixed rate.

WDYT?

Reply:
I have added said check and updated the comment to be more clear. I have also added a test covering this scenario.